### DISCUSSION TITLE
Monitoring writers and readers with plots (series Ids are different for every run)

### DISCUSSION THREAD

**Discussion by KansaiUser** [eProsima/Fast-DDS-monitor#184] [Q&A]
I have managed to have compile a simple publisher subscriber program with CMake option `-DFASTDDS_STATISTICS=ON`

I also installed fast-DDS Monitor.
After exporting statistics topics with
```
export FASTDDS_STATISTICS="HISTORY_LATENCY_TOPIC;PUBLICATION_THROUGHPUT_TOPIC;ACKNACK_COUNT_TOPIC;;DISCOVERY_TOPIC;PHYSICAL_DATA_TOPIC;SUBSCRIPTION_THROUGHPUT_TOPIC"
```
I run the subscriber and publisher 
and I run the monitor and I could see the Participant_sub and Participant_pub.
I also could add a plot of Latency

I have the following questions:

- The main one: Every time I run the publisher and subscriber, new entities appear in the DDS Entities explorer with their datareaders and datawriters with a different ID. This would not be a problem but when I want to make a plot, I have to add series indicating those IDS. 
The problem is that I don't have those IDs _before_ running the entities. So I can not design the plot I want before running the entities. This implies that I have to design the plots in a hurry after the entities started running. This for a plot is problematic but for several plots becomes impossible and error prone. 

How can I prepare the plots I want with the series I want to monitor _before_ I run my application?

Update:
It seems that even though the plot's series indicate some ID numbers for readers and writers, when I run them again (and they get different numbers, they still get plotted. Why is this so?

- The monitoring seems to be going well but when I run the program I got an Error (that does not stop the entity though)
It says
```
2023-05-04 19:49:56.377 [STATISTICS_DOMAIN_PARTICIPANT_Error] Topic   is not a valid statistic topic name/alias -> Function enable_statistics_builtin_datawriters
```
What does this mean?

Thanks

**Comment by jparisu** [comment]
This is a question rather than an issue. I transfer it to discussions.

**Comment by jparisu** [comment]
Hi @KansaiUser 

The information regarding your use case can be found [here](<https://fast-dds-monitor.readthedocs.io/en/latest/rst/user_manual/chart_panel/chart_panel.html#source-entity-id>).

Let's take it step by step:

---

## Unique ids

1. There are 2 different Ids in the Fast DDS Monitor. 
    1. One is the DDS GUID (01.0f.00...). This is random as required by the standard.
    2. The other is the unique Id that the Statistics Backend gives to each entity. This one is generated each time an entity is found.

The DDS GUID could be set in the origin entities by configuring by XML. However, this is not advisable for non expert users as an entity with the same GUID is considered as the same entity, and could create discovery or communication problems.
The Backend Id cannot be changed. 

---

## Plot information

When plotting data, it must be chosen a source entity and depending on the data kind, a target entity as well.
For different Data Kinds, the entity required is different. 
The relation between entities and Data Kinds can be found [here](https://fast-dds-statistics-backend.readthedocs.io/en/latest/rst/types/data_kind.html).

However, Fast DDS Monitor adds a feature that allows to **join** several entities that are releated with each other.
These connections could be seen [here](https://fast-dds-monitor.readthedocs.io/en/latest/rst/getting_started/entities.html#entities).

Let's say, for example, that we want the Latency (that is measured from a DataWriter to a DataReader) of the writer `0` to the reader `1`.
We will create the plot, and use id 0 as source and id 1 as target.
However, if you run a different writer and/or reader, these new entities will not be 0 and 1, but some other new Ids.
Thus, in order to plot all the data that goes from **any** writer to **any** reader in a specific topic, you can set as source the id of the DDS topic, and as target the same Id of the same topic. This will get all active writers and readers in such topic and calculate the latency between them (data will be grouped depending on the Statistics Kind).

---

## Solution

Use some way of grouping your writers and readers so the plot works when you run new instances.
Topic would be a good option. 
Also Domain or Host would be useful.
